File: Codes from the AI-assistants/fasta/Fasta-ds-p3-normal.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define IM 139968
#define IA 3877
#define IC 29573
#define LINE_LENGTH 60

static unsigned int seed = 42;

double random_number(double max) {
    seed = (seed * IA + IC) % IM;
    return max * seed / IM;
}

void make_cumulative(char *chars, double *probs, int count) {
    double cumulative = 0.0;
    for (int i = 0; i < count; i++) {
        cumulative += probs[i];
        probs[i] = cumulative;
    }
}

char select_random(char *chars, double *probs, int count) {
    double r = random_number(1.0);
    
    // Linear search
    for (int i = 0; i < count; i++) {
        if (r <= probs[i]) {
            return chars[i];
        }
    }
    return chars[count - 1];
}

void generate_repeat_fasta(const char *seq, int n) {
    int seq_len = strlen(seq);
    int pos = 0;
    char buffer[LINE_LENGTH + 1];
    
    while (n > 0) {
        int chunk = n < LINE_LENGTH ? n : LINE_LENGTH;
        if (seq_len - pos < chunk) {
            chunk = seq_len - pos;
        }
        
        strncpy(buffer, seq + pos, chunk);
        buffer[chunk] = '\0';
        printf("%s\n", buffer);
        
        pos += chunk;
        if (pos >= seq_len) {
            pos = 0;
        }
        n -= chunk;
    }
}

void generate_random_fasta(char *chars, double *probs, int count, int n) {
    make_cumulative(chars, probs, count);
    char buffer[LINE_LENGTH + 1];
    
    while (n > 0) {
        int chunk = n < LINE_LENGTH ? n : LINE_LENGTH;
        
        for (int i = 0; i < chunk; i++) {
            buffer[i] = select_random(chars, probs, count);
        }
        buffer[chunk] = '\0';
        printf("%s\n", buffer);
        
        n -= chunk;
    }
}

int main(int argc, char *argv[]) {
    int n = 1000;
    if (argc > 1) {
        n = atoi(argv[1]);
    }
    
    const char *alu =
        "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG"
        "GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA"
        "CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT"
        "ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA"
        "GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG"
        "AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC"
        "AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA";
    
    // IUB table
    char iub_chars[] = "acgtBDHKMNRSVWY";
    double iub_probs[] = {
        0.27, 0.12, 0.12, 0.27,
        0.02, 0.02, 0.02, 0.02,
        0.02, 0.02, 0.02, 0.02,
        0.02, 0.02, 0.02
    };
    
    // Homo sapiens table
    char homosapiens_chars[] = "acgt";
    double homosapiens_probs[] = {
        0.3029549426680,
        0.1979883004921,
        0.1975473066391,
        0.3015094502008
    };
    
    printf(">ONE Homo sapiens alu\n");
    generate_repeat_fasta(alu, n * 2);
    
    printf(">TWO IUB ambiguity codes\n");
    seed = 42;
    generate_random_fasta(iub_chars, iub_probs, 15, n * 3);
    
    printf(">THREE Homo sapiens frequency\n");
    seed = 42;
    generate_random_fasta(homosapiens_chars, homosapiens_probs, 4, n * 5);
    
    return 0;
}

File: Codes from the AI-assistants/fasta/Fasta-ds-p3-normal.py
import sys
import bisect

# Constants for LCG
IM = 139968
IA = 3877
IC = 29573
SEED = 42

# LCG implementation as specified
def random_number(max_val):
    global SEED
    SEED = (SEED * IA + IC) % IM
    return max_val * SEED / IM

def make_cumulative(table):
    """Convert probability table to cumulative probabilities"""
    # Create list of (nucleotide, probability)
    items = list(table.items())
    
    # Calculate cumulative probabilities (naïve approach)
    cumulative = []
    current = 0.0
    for char, prob in items:
        current += prob
        cumulative.append((char, current))
    
    return cumulative

def select_random(cumulative):
    """Select nucleotide using cumulative probabilities with linear search"""
    r = random_number(1.0)
    
    # Linear search as specified (not binary)
    for char, cum_prob in cumulative:
        if r <= cum_prob:
            return char
    return cumulative[-1][0]  # Fallback

def generate_repeat_fasta(seq, n):
    """Generate DNA by copying from given sequence"""
    out = []
    seq_len = len(seq)
    pos = 0
    
    while n > 0:
        chunk = min(n, seq_len - pos, 60)
        out.append(seq[pos:pos + chunk])
        pos = (pos + chunk) % seq_len
        n -= chunk
        if n > 0 and chunk == 60:
            out.append('\n')
    
    sys.stdout.write(''.join(out) + '\n')

def generate_random_fasta(table, n):
    """Generate DNA using weighted random selection"""
    # Create cumulative probabilities
    cumulative = make_cumulative(table)
    
    out = []
    while n > 0:
        chunk = min(n, 60)
        for i in range(chunk):
            out.append(select_random(cumulative))
        out.append('\n')
        n -= chunk
    
    sys.stdout.write(''.join(out))

def main():
    n = 1000
    if len(sys.argv) > 1:
        n = int(sys.argv[1])
    
    # Reset seed for each test
    global SEED
    SEED = 42
    
    # ALU sequence
    alu = (
        'GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG'
        'GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA'
        'CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT'
        'ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA'
        'GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG'
        'AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC'
        'AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA'
    )
    
    # Probability tables
    iub = {
        'a': 0.27, 'c': 0.12, 'g': 0.12, 't': 0.27,
        'B': 0.02, 'D': 0.02, 'H': 0.02, 'K': 0.02,
        'M': 0.02, 'N': 0.02, 'R': 0.02, 'S': 0.02,
        'V': 0.02, 'W': 0.02, 'Y': 0.02
    }
    
    homosapiens = {
        'a': 0.3029549426680,
        'c': 0.1979883004921,
        'g': 0.1975473066391,
        't': 0.3015094502008
    }
    
    # Generate sequences
    sys.stdout.write(f'>ONE Homo sapiens alu\n')
    generate_repeat_fasta(alu, n * 2)
    
    sys.stdout.write(f'>TWO IUB ambiguity codes\n')
    SEED = 42  # Reset seed
    generate_random_fasta(iub, n * 3)
    
    sys.stdout.write(f'>THREE Homo sapiens frequency\n')
    SEED = 42  # Reset seed
    generate_random_fasta(homosapiens, n * 5)

if __name__ == '__main__':
    main()